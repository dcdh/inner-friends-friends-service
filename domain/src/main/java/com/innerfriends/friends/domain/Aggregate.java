package com.innerfriends.friends.domain;

import java.util.Objects;
import java.util.concurrent.Callable;

public abstract class Aggregate {

    private Version version;

    public Aggregate(final Version version) {
        this.version = version;
    }

    protected void apply(final Callable apply) {
        try {
            apply.call();
        } catch (final Exception exception) {
            throw new FailedToMutateAggregateException(exception);
        }
        this.version = this.version.increment();
    }

    public Version version() {
        return version;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (!(o instanceof Aggregate)) return false;
        final Aggregate aggregate = (Aggregate) o;
        return Objects.equals(version, aggregate.version);
    }

    @Override
    public int hashCode() {
        return Objects.hash(version);
    }

    @Override
    public String toString() {
        return "Aggregate{" +
                "version=" + version +
                '}';
    }
}
