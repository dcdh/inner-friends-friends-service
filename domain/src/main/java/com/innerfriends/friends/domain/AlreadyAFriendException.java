package com.innerfriends.friends.domain;

public class AlreadyAFriendException extends RuntimeException {
}
