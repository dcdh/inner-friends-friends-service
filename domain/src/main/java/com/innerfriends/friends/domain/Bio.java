package com.innerfriends.friends.domain;

import java.util.Objects;

public final class Bio {

    private final String content;

    public Bio() {
        this("");
    }

    public Bio(final String content) {
        this.content = Objects.requireNonNull(content);
        if (content.length() >= 500) {
            throw new IllegalStateException();
        }
    }

    public String content() {
        return content;
    }

    public boolean hasBio() {
        return content.length() > 0;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (!(o instanceof Bio)) return false;
        final Bio bio1 = (Bio) o;
        return Objects.equals(content, bio1.content);
    }

    @Override
    public int hashCode() {
        return Objects.hash(content);
    }
}
