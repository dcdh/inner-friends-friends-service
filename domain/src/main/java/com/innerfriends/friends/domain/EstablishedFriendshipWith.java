package com.innerfriends.friends.domain;

import java.util.Objects;

public final class EstablishedFriendshipWith {

    private final InFriendshipWithId inFriendshipWithId;

    public EstablishedFriendshipWith(final InFriendshipWithId inFriendshipWithId) {
        this.inFriendshipWithId = Objects.requireNonNull(inFriendshipWithId);
    }

    public InFriendshipWithId inFriendshipWithId() {
        return inFriendshipWithId;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (!(o instanceof EstablishedFriendshipWith)) return false;
        final EstablishedFriendshipWith that = (EstablishedFriendshipWith) o;
        return Objects.equals(inFriendshipWithId, that.inFriendshipWithId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(inFriendshipWithId);
    }

    @Override
    public String toString() {
        return "EstablishedFriendshipWith{" +
                "inFriendshipWithId=" + inFriendshipWithId +
                '}';
    }

}
