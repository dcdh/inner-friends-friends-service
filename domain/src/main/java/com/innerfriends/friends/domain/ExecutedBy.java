package com.innerfriends.friends.domain;

import java.util.Objects;

public final class ExecutedBy implements Pseudo {

    private final PseudoId pseudoId;

    public ExecutedBy(final PseudoId pseudoId) {
        this.pseudoId = Objects.requireNonNull(pseudoId);
    }

    public ExecutedBy(final Pseudo pseudo) {
        this(Objects.requireNonNull(pseudo)
                .pseudoId());
    }

    public ExecutedBy(final String pseudo) {
        this(new PseudoId(pseudo));
    }

    @Override
    public PseudoId pseudoId() {
        return pseudoId;
    }

    public String pseudo() {
        return pseudoId.pseudo();
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (!(o instanceof ExecutedBy)) return false;
        final ExecutedBy that = (ExecutedBy) o;
        return Objects.equals(pseudoId, that.pseudoId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(pseudoId);
    }

    @Override
    public String toString() {
        return "ExecutedBy{" +
                "pseudoId=" + pseudoId +
                '}';
    }
}
