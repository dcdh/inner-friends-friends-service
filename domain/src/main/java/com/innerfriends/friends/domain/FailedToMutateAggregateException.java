package com.innerfriends.friends.domain;

public class FailedToMutateAggregateException extends RuntimeException {

    public FailedToMutateAggregateException(final Throwable cause) {
        super(cause);
    }
}
