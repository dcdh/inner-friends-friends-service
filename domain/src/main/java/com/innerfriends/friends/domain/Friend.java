package com.innerfriends.friends.domain;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

public final class Friend extends Aggregate {

    public static final InFriendshipWithId DAM_DAM_DEO = new InFriendshipWithId("DamDamDeo");

    private final FriendId friendId;
    private final List<InFriendshipWithId> friendshipsWith;
    private Bio bio;

    public Friend(final FriendId friendId) {
        this(friendId, new ArrayList<>(List.of(DAM_DAM_DEO)),
                new Bio(), new Version(0l));
    }

    public Friend(final FriendId friendId,
                  final List<InFriendshipWithId> friendshipsWith,
                  final Bio bio,
                  final Version version) {
        super(version);
        this.friendId = Objects.requireNonNull(friendId);
        this.friendshipsWith = Objects.requireNonNull(friendshipsWith);
        this.bio = Objects.requireNonNull(bio);
    }

    public Friend writeBio(final Bio bio, final ExecutedBy executedBy) {
        if (!executedBy.pseudoId().equals(friendId.pseudoId())) {
            throw new IllegalStateException("The one doing the action must be the friend");
        }
        this.apply(() -> this.bio = bio);
        return this;
    }

    public Friend establishAFriendship(final FriendId friendIdToEstablishAFriendshipWith, final ExecutedBy executedBy) {
        if (!executedBy.pseudoId().equals(friendId.pseudoId())) {
            throw new IllegalStateException("The one doing the action must be the friend");
        }
        this.establishAFriendship(friendIdToEstablishAFriendshipWith);
        return this;
    }

    public Friend establishAFriendship(final FriendId friendIdToEstablishAFriendshipWith) {
        if (friendId.pseudoId().equals(friendIdToEstablishAFriendshipWith.pseudoId())) {
            throw new IllegalStateException("You could not add yourself as a friend");
        }
        if (DAM_DAM_DEO.pseudoId().equals(friendIdToEstablishAFriendshipWith.pseudoId())) {
            return this;
        }
        if (friendshipsWith.contains(new InFriendshipWithId(friendIdToEstablishAFriendshipWith))) {
            throw new AlreadyAFriendException();
        }
        this.apply(() -> {
            friendshipsWith.add(new InFriendshipWithId(friendIdToEstablishAFriendshipWith));
            return null;
        });
        return this;
    }

    public FriendId friendId() {
        return friendId;
    }

    public List<InFriendshipWithId> inFriendshipsWith() {
        return Collections.unmodifiableList(friendshipsWith);
    }

    public Bio bio() {
        return bio;
    }

    public Integer nbOfFriends() {
        return friendshipsWith.size();
    }

    public boolean hasBio() {
        return bio.hasBio();
    }

    public boolean isInFriendshipWith(final InFriendshipWithId inFriendshipWithId) {
        return friendshipsWith.contains(inFriendshipWithId);
    }

    public InFriendshipWithId lastEstablishedFriendshipWith() {
        return friendshipsWith.stream().reduce((first, second) -> second).get();
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (!(o instanceof Friend)) return false;
        if (!super.equals(o)) return false;
        final Friend friend = (Friend) o;
        return Objects.equals(friendId, friend.friendId) &&
                Objects.equals(friendshipsWith, friend.friendshipsWith) &&
                Objects.equals(bio, friend.bio);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), friendId, friendshipsWith, bio);
    }

    @Override
    public String toString() {
        return "Friend{" +
                "friendId=" + friendId +
                ", friendshipsWith=" + friendshipsWith +
                ", bio=" + bio +
                "} " + super.toString();
    }
}
