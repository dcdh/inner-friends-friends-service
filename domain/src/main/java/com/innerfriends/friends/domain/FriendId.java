package com.innerfriends.friends.domain;

import java.util.Objects;

public final class FriendId implements Pseudo {

    private final PseudoId pseudoId;

    public FriendId(final PseudoId pseudoId) {
        this.pseudoId = Objects.requireNonNull(pseudoId);
    }

    public FriendId(final Pseudo pseudo) {
        this(Objects.requireNonNull(pseudo)
                .pseudoId());
    }

    public FriendId(final String pseudo) {
        this(new PseudoId(pseudo));
    }

    @Override
    public PseudoId pseudoId() {
        return pseudoId;
    }

    public String pseudo() {
        return pseudoId.pseudo();
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (!(o instanceof FriendId)) return false;
        final FriendId friendId = (FriendId) o;
        return Objects.equals(pseudoId, friendId.pseudoId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(pseudoId);
    }

    @Override
    public String toString() {
        return "FriendId{" +
                "pseudoId=" + pseudoId +
                '}';
    }
}
