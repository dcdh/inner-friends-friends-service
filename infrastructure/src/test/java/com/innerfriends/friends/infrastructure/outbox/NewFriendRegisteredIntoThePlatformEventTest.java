package com.innerfriends.friends.infrastructure.outbox;

import com.innerfriends.friends.domain.Friend;
import com.innerfriends.friends.domain.FriendId;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.Instant;

import static org.approvaltests.Approvals.verifyJson;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.doReturn;

@ExtendWith(MockitoExtension.class)
public class NewFriendRegisteredIntoThePlatformEventTest extends EventTest {

    @Test
    public void should_return_expected_event() {
        // Given
        final Friend mario = new Friend(new FriendId("Mario"));
        doReturn(Instant.ofEpochSecond(1)).when(instantProvider).now();

        // When
        final NewFriendRegisteredIntoThePlatformEvent newFriendRegisteredIntoThePlatformEvent = NewFriendRegisteredIntoThePlatformEvent
                .of(mario, objectMapper, instantProvider);

        // Then
        assertThat(newFriendRegisteredIntoThePlatformEvent.getAggregateId()).isEqualTo("Mario");
        assertThat(newFriendRegisteredIntoThePlatformEvent.getAggregateType()).isEqualTo("Friend");
        assertThat(newFriendRegisteredIntoThePlatformEvent.getTimestamp()).isEqualTo(Instant.ofEpochSecond(1));
        assertThat(newFriendRegisteredIntoThePlatformEvent.getType()).isEqualTo("NewFriendRegisteredIntoThePlatform");
        verifyJson(newFriendRegisteredIntoThePlatformEvent.getPayload().toString());
    }
}
