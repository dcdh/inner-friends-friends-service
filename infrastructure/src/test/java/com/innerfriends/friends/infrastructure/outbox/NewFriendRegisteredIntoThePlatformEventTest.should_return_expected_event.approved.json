{
  "friendId": "Mario",
  "version": 0,
  "inFriendshipsWith": [
    "DamDamDeo"
  ]
}